File: core/src/main/java/fund/cyber/markets/common/CircularQueue.java
package fund.cyber.markets.common;

public class CircularQueue<T> {

    private final int maxSize;

    private T[] elements;
    private int start = 0;
    private int end = 0;

    private boolean isFull = false;

    public CircularQueue(int maxSize) {
        this.maxSize = maxSize;
        elements = (T[]) new Object[maxSize];

    }


    public void addNext(T element) {

        elements[end++] = element;

        if (isFull) start++;
        if (start >= maxSize) start = 0;
        if (end >= maxSize) end = 0;
        if (end == start) isFull = true;
    }

    public T[] getElements() {
        return elements;
    }

    public T getElement(int indexInQueue) {
        if (start + indexInQueue < 10) {
            return elements[start+indexInQueue];
        } else {
            return elements[start+indexInQueue-10];
        }
    }

}

File: core/src/main/java/fund/cyber/markets/common/CircularQueue.kt
package fund.cyber.markets.common

/**
 * Queue with limited size, which overrides elements like it a circle.
 *
 * @author Ibragimov Ruslan
 * @since 0.2.4
 */
class CircularQueueKotlin<T>(
    private val maxSize: Int
) {
    private val _elements = arrayOfNulls<Any>(maxSize)
    private var index = 0

    fun addNext(element: T) {
        _elements[index++] = element
        if (index == maxSize) index = 0
    }

    val elements: Array<T?>
        get() = _elements.clone() as Array<T?>
}